test/ecto/query/planner_test.exs

```diff
@@ -848,11 +848,23 @@ defmodule Ecto.Query.PlannerTest do
     assert_raise Ecto.QueryError, ~r/could not find named binding `as\(:posts\)`/, fn ->
       from(Post, where: as(^as).code == ^123) |> normalize()
     end
+  end
+
+  test "normalize: creating dynamic bindings with as" do
+    as = [:posts]
+
+    query = from(Post, as: as, where: as(^as).code == ^123) |> normalize()
+    assert Macro.to_string(hd(query.wheres).expr) == "&0.code() == ^0"
 
-    assert_raise Ecto.Query.CompileError, ~r/expected atom in as\/1/, fn ->
-      as = "posts"
+    query = from(Post, as: as, where: field(as(^as), :code) == ^123) |> normalize()
```

Comment:
I am thinking we should also use `as: ^as`. Seeing the "naked variable" is not very clear on intent because naked variables in queries usually refer to a table? WDYT?

Reply:
Yeah, that makes a lot of sense, Ill make that change.